maven-core/src/main/java/org/apache/maven/internal/transformation/impl/TransformedArtifact.java

```diff
@@ -0,0 +1,148 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.internal.transformation.impl;
+
+import javax.xml.stream.XMLStreamException;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import org.apache.maven.artifact.DefaultArtifact;
+import org.apache.maven.internal.transformation.TransformationFailedException;
+import org.apache.maven.model.building.ModelBuildingException;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.aether.RepositorySystemSession;
+
+/**
+ * Transformed artifact is derived with some transformation from source artifact.
+ *
+ * @since TBD
+ */
+class TransformedArtifact extends DefaultArtifact {
+
+    private static final int SHA1_BUFFER_SIZE = 8192;
+    private final DefaultConsumerPomArtifactTransformer defaultConsumerPomArtifactTransformer;
+    private final MavenProject project;
+    private final Supplier<Path> sourcePathProvider;
+    private final Path target;
+    private final RepositorySystemSession session;
+    private final AtomicReference<String> sourceState;
+
+    TransformedArtifact(
+            DefaultConsumerPomArtifactTransformer defaultConsumerPomArtifactTransformer,
+            MavenProject project,
+            Path target,
+            RepositorySystemSession session,
+            org.apache.maven.artifact.Artifact source,
+            Supplier<Path> sourcePathProvider,
+            String classifier,
+            String extension) {
+        super(
+                source.getGroupId(),
+                source.getArtifactId(),
+                source.getVersionRange(),
+                source.getScope(),
+                extension,
+                classifier,
+                new TransformedArtifactHandler(
+                        classifier, extension, source.getArtifactHandler().getPackaging()));
+        this.defaultConsumerPomArtifactTransformer = defaultConsumerPomArtifactTransformer;
+        this.project = project;
+        this.target = target;
+        this.session = session;
+        this.sourcePathProvider = sourcePathProvider;
+        this.sourceState = new AtomicReference<>(null);
+    }
+
+    @Override
+    public boolean isResolved() {
+        return getFile() != null;
+    }
+
+    @Override
+    public void setFile(File file) {
+        throw new IllegalStateException("transformed artifact file cannot be set");
```

Comment:
UnsupportedOperationException feels more correct here, but why is this even impossible? Why can't a transformed artifact have a file?

Reply:
Fixed the exception thrown.